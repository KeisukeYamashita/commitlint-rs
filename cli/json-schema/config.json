{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "title": "Config",
  "description": "Config represents the configuration of commitlint.",
  "type": "object",
  "properties": {
    "rules": {
      "description": "Rules represents the rules of commitlint.",
      "$ref": "#/$defs/Rules"
    }
  },
  "required": [
    "rules"
  ],
  "$defs": {
    "BodyEmpty": {
      "description": "BodyEmpty represents the body-empty rule.",
      "type": "object",
      "properties": {
        "level": {
          "description": "Level represents the level of the rule.",
          "anyOf": [
            {
              "$ref": "#/$defs/Level"
            },
            {
              "type": "null"
            }
          ]
        }
      }
    },
    "BodyMaxLength": {
      "description": "BodyMaxLength represents the body-max-length rule.",
      "type": "object",
      "properties": {
        "length": {
          "description": "Length represents the maximum length of the length_of_what.",
          "type": "integer",
          "format": "uint",
          "minimum": 0
        },
        "level": {
          "description": "Level represents the level of the rule.",
          "anyOf": [
            {
              "$ref": "#/$defs/Level"
            },
            {
              "type": "null"
            }
          ]
        }
      },
      "required": [
        "length"
      ]
    },
    "DescriptionEmpty": {
      "description": "DescriptionEmpty represents the description-empty rule",
      "type": "object",
      "properties": {
        "level": {
          "description": "Level represents the level of the rule.",
          "anyOf": [
            {
              "$ref": "#/$defs/Level"
            },
            {
              "type": "null"
            }
          ]
        }
      }
    },
    "DescriptionFormat": {
      "description": "DescriptionFormat represents the description-format rule.",
      "type": "object",
      "properties": {
        "format": {
          "description": "Format represents the format of the length_of_what.",
          "type": [
            "string",
            "null"
          ]
        },
        "level": {
          "description": "Level represents the level of the rule.",
          "anyOf": [
            {
              "$ref": "#/$defs/Level"
            },
            {
              "type": "null"
            }
          ]
        }
      }
    },
    "DescriptionMaxLength": {
      "description": "DescriptionMaxLength represents the description-max-length rule.",
      "type": "object",
      "properties": {
        "length": {
          "description": "Length represents the maximum length of the length_of_what.",
          "type": "integer",
          "format": "uint",
          "minimum": 0
        },
        "level": {
          "description": "Level represents the level of the rule.",
          "anyOf": [
            {
              "$ref": "#/$defs/Level"
            },
            {
              "type": "null"
            }
          ]
        }
      },
      "required": [
        "length"
      ]
    },
    "FootersEmpty": {
      "description": "FootersEmpty represents the footer-empty rule",
      "type": "object",
      "properties": {
        "level": {
          "description": "Level represents the level of the rule.",
          "anyOf": [
            {
              "$ref": "#/$defs/Level"
            },
            {
              "type": "null"
            }
          ]
        }
      }
    },
    "Level": {
      "description": "Level represents the level of a rule.",
      "type": "string",
      "enum": [
        "error",
        "ignore",
        "warning"
      ]
    },
    "Rules": {
      "description": "Rules represents the rules of commitlint.\n See: https://commitlint.js.org/reference/rules.html",
      "type": "object",
      "properties": {
        "type": {
          "anyOf": [
            {
              "$ref": "#/$defs/Type"
            },
            {
              "type": "null"
            }
          ]
        },
        "body-empty": {
          "anyOf": [
            {
              "$ref": "#/$defs/BodyEmpty"
            },
            {
              "type": "null"
            }
          ]
        },
        "body-max-length": {
          "anyOf": [
            {
              "$ref": "#/$defs/BodyMaxLength"
            },
            {
              "type": "null"
            }
          ]
        },
        "description-empty": {
          "anyOf": [
            {
              "$ref": "#/$defs/DescriptionEmpty"
            },
            {
              "type": "null"
            }
          ]
        },
        "description-format": {
          "anyOf": [
            {
              "$ref": "#/$defs/DescriptionFormat"
            },
            {
              "type": "null"
            }
          ]
        },
        "description-max-length": {
          "anyOf": [
            {
              "$ref": "#/$defs/DescriptionMaxLength"
            },
            {
              "type": "null"
            }
          ]
        },
        "footers-empty": {
          "anyOf": [
            {
              "$ref": "#/$defs/FootersEmpty"
            },
            {
              "type": "null"
            }
          ]
        },
        "scope": {
          "anyOf": [
            {
              "$ref": "#/$defs/Scope"
            },
            {
              "type": "null"
            }
          ]
        },
        "scope-empty": {
          "anyOf": [
            {
              "$ref": "#/$defs/ScopeEmpty"
            },
            {
              "type": "null"
            }
          ]
        },
        "scope-format": {
          "anyOf": [
            {
              "$ref": "#/$defs/ScopeFormat"
            },
            {
              "type": "null"
            }
          ]
        },
        "scope-max-length": {
          "anyOf": [
            {
              "$ref": "#/$defs/ScopeMaxLength"
            },
            {
              "type": "null"
            }
          ]
        },
        "subject-empty": {
          "anyOf": [
            {
              "$ref": "#/$defs/SubjectEmpty"
            },
            {
              "type": "null"
            }
          ]
        },
        "type-empty": {
          "anyOf": [
            {
              "$ref": "#/$defs/TypeEmpty"
            },
            {
              "type": "null"
            }
          ]
        },
        "type-format": {
          "anyOf": [
            {
              "$ref": "#/$defs/TypeFormat"
            },
            {
              "type": "null"
            }
          ]
        },
        "type-max-length": {
          "anyOf": [
            {
              "$ref": "#/$defs/TypeMaxLength"
            },
            {
              "type": "null"
            }
          ]
        }
      }
    },
    "Scope": {
      "description": "Scope represents the subject rule.",
      "type": "object",
      "properties": {
        "level": {
          "description": "Level represents the level of the rule.",
          "anyOf": [
            {
              "$ref": "#/$defs/Level"
            },
            {
              "type": "null"
            }
          ]
        },
        "optional": {
          "description": "Optional scope. If true, even if the scope is not present, it is allowed.",
          "type": "boolean"
        },
        "options": {
          "description": "Options represents the options of the rule. If the option is empty, it means that no options_what is allowed.",
          "type": "array",
          "items": {
            "type": "string"
          }
        }
      },
      "required": [
        "optional",
        "options"
      ]
    },
    "ScopeEmpty": {
      "description": "FootersEmpty represents the subject-empty rule",
      "type": "object",
      "properties": {
        "level": {
          "description": "Level represents the level of the rule.",
          "anyOf": [
            {
              "$ref": "#/$defs/Level"
            },
            {
              "type": "null"
            }
          ]
        }
      }
    },
    "ScopeFormat": {
      "description": "ScopeFormat represents the scope-format rule.",
      "type": "object",
      "properties": {
        "format": {
          "description": "Format represents the format of the length_of_what.",
          "type": [
            "string",
            "null"
          ]
        },
        "level": {
          "description": "Level represents the level of the rule.",
          "anyOf": [
            {
              "$ref": "#/$defs/Level"
            },
            {
              "type": "null"
            }
          ]
        }
      }
    },
    "ScopeMaxLength": {
      "description": "ScopeMaxLength represents the scope-max-length rule.",
      "type": "object",
      "properties": {
        "length": {
          "description": "Length represents the maximum length of the length_of_what.",
          "type": "integer",
          "format": "uint",
          "minimum": 0
        },
        "level": {
          "description": "Level represents the level of the rule.",
          "anyOf": [
            {
              "$ref": "#/$defs/Level"
            },
            {
              "type": "null"
            }
          ]
        }
      },
      "required": [
        "length"
      ]
    },
    "SubjectEmpty": {
      "description": "SubjectEmpty represents the subject-empty rule.",
      "type": "object",
      "properties": {
        "level": {
          "description": "Level represents the level of the rule.",
          "anyOf": [
            {
              "$ref": "#/$defs/Level"
            },
            {
              "type": "null"
            }
          ]
        }
      }
    },
    "Type": {
      "description": "Type represents the type rule.",
      "type": "object",
      "properties": {
        "level": {
          "description": "Level represents the level of the rule.",
          "anyOf": [
            {
              "$ref": "#/$defs/Level"
            },
            {
              "type": "null"
            }
          ]
        },
        "options": {
          "description": "Options represents the options of the rule. If the option is empty, it means that no options_what is allowed.",
          "type": "array",
          "items": {
            "type": "string"
          }
        }
      },
      "required": [
        "options"
      ]
    },
    "TypeEmpty": {
      "description": "TypeEmpty represents the type-empty rule.",
      "type": "object",
      "properties": {
        "level": {
          "description": "Level represents the level of the rule.",
          "anyOf": [
            {
              "$ref": "#/$defs/Level"
            },
            {
              "type": "null"
            }
          ]
        }
      }
    },
    "TypeFormat": {
      "description": "TypeFormat represents the type-format rule.",
      "type": "object",
      "properties": {
        "format": {
          "description": "Format represents the format of the length_of_what.",
          "type": [
            "string",
            "null"
          ]
        },
        "level": {
          "description": "Level represents the level of the rule.",
          "anyOf": [
            {
              "$ref": "#/$defs/Level"
            },
            {
              "type": "null"
            }
          ]
        }
      }
    },
    "TypeMaxLength": {
      "description": "TypeMaxLength represents the type-max-length rule.",
      "type": "object",
      "properties": {
        "length": {
          "description": "Length represents the maximum length of the length_of_what.",
          "type": "integer",
          "format": "uint",
          "minimum": 0
        },
        "level": {
          "description": "Level represents the level of the rule.",
          "anyOf": [
            {
              "$ref": "#/$defs/Level"
            },
            {
              "type": "null"
            }
          ]
        }
      },
      "required": [
        "length"
      ]
    }
  }
}