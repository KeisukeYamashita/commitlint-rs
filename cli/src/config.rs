use serde::{Deserialize, Serialize};
use std::fmt;
use std::{fs, path::PathBuf};

use crate::rule::Rules;

/// Default Root config file path to search for.
const DEFAULT_CONFIG_ROOT: &str = ".";

/// Default commitlintrc configuration files
/// If the user didn't specify a configuration file with -c or --config argument,
/// we will try to find one of these files in the current directory.
const DEFAULT_CONFIG_FILE: [&str; 4] = [
    ".commitlintrc",
    ".commitlintrc.json",
    ".commitlintrc.yaml",
    ".commitlintrc.yml",
];

/// Config represents the configuration of commitlint.
#[derive(Clone, Debug, Default, Deserialize, Serialize)]
#[cfg_attr(feature = "schemars", derive(schemars::JsonSchema))]
pub struct Config {
    /// Rules represents the rules of commitlint.
    pub rules: Rules,
}

impl fmt::Display for Config {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = serde_yaml::to_string(&self).unwrap();
        write!(f, "{}", s)
    }
}

/// Load configuration from the specified path.
pub async fn load(path: Option<PathBuf>) -> Result<Config, String> {
    let config_file = match &path {
        Some(p) => Some(p.clone()),
        None => find_config_file(PathBuf::from(DEFAULT_CONFIG_ROOT)),
    };

    match (config_file, path) {
        // If the file was specified and found, load it.
        (Some(p), _) => load_config_file(p).await,
        // If the file was not specified and not found, return default config.
        (None, None) => Ok(Config::default()),
        // If the was explicitly specified but not found, return an error.
        (None, Some(p)) => Err(format!("Configuration file not found in {}", p.display())),
    }
}

/// Find configuration file in the specified path.
/// Note that the first file found will be returned.
pub fn find_config_file(path: PathBuf) -> Option<PathBuf> {
    let mut path = path;
    for file in DEFAULT_CONFIG_FILE.iter() {
        path.push(file);
        if path.exists() {
            return Some(path);
        }
        path.pop();
    }

    None
}

/// Load config file from the specified path.
pub async fn load_config_file(path: PathBuf) -> Result<Config, String> {
    if !path.exists() {
        return Err(format!(
            "Configuration file not found in {}",
            path.display()
        ));
    }

    match path.extension() {
        Some(ext) => match ext.to_str() {
            Some("json") => load_json_config_file(path).await,
            Some("yaml") | Some("yml") => load_yaml_config_file(path).await,
            _ => load_unknown_config_file(path).await,
        },
        None => Err(format!(
            "Unsupported configuration file format: {}",
            path.display()
        )),
    }
}

/// Load JSON config file from the specified path.
async fn load_json_config_file(path: PathBuf) -> Result<Config, String> {
    let text = fs::read_to_string(path).unwrap();

    match serde_json::from_str::<Config>(&text) {
        Ok(config) => Ok(config),
        Err(err) => Err(format!("Failed to parse configuration file: {}", err)),
    }
}

/// Load YAML config file from the specified path.
async fn load_yaml_config_file(path: PathBuf) -> Result<Config, String> {
    let text = fs::read_to_string(path).unwrap();

    match serde_yaml::from_str::<Config>(&text) {
        Ok(config) => Ok(config),
        Err(err) => Err(format!("Failed to parse configuration file: {}", err)),
    }
}

/// Try to load configuration file from the specified path.
/// First try to load it as JSON, then as YAML.
/// If both fail, return an error.
async fn load_unknown_config_file(path: PathBuf) -> Result<Config, String> {
    let text = fs::read_to_string(path.clone()).unwrap();

    if let Ok(config) = serde_json::from_str::<Config>(&text) {
        return Ok(config);
    }

    if let Ok(config) = serde_yaml::from_str::<Config>(&text) {
        return Ok(config);
    }

    Err(format!(
        "Failed to parse configuration file: {}",
        path.display()
    ))
}
