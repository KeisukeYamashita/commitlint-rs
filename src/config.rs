use std::{fmt::Error, path::PathBuf};

/// Default commitlintrc configuration files
/// If the user didn't specify a configuration file with -g or --config argument,
/// we will try to find one of these files in the current directory.
const DEFAULT_CONFIG_FILE: [&str; 4] = [
    ".commitlintrc",
    ".commitlintrc.json",
    ".commitlintrc.yaml",
    ".commitlintrc.yml",
];

/// Config represents the configuration of commitlint.
#[derive(Debug)]
pub struct Config {}

/// Default configuration if no configuration file is found.
pub fn default_config() -> Config {
    Config {}
}

/// Load configuration from the specified path.
pub async fn load(path: PathBuf) -> Result<Config, Error> {
    let config_file = find_config_file(path);

    if config_file.is_none() {
        return Ok(default_config());
    }

    Ok(Config {})
}

/// Find configuration file in the specified path.
/// Note that the first file found will be returned.
pub fn find_config_file(path: PathBuf) -> Option<PathBuf> {
    let mut path = path;
    for file in DEFAULT_CONFIG_FILE.iter() {
        path.push(file);
        if path.exists() {
            return Some(path);
        }
    }

    None
}
