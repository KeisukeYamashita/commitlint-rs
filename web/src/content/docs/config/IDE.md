---
title: IDE
description: Guide how to setup your IDE to work with commitlint
---

Commitlint offers schema by supporting [JSON schema](https://json-schema.org/) so that you can configure your IDE to work with Commitlint and have better developer experience.

:::tip

If you want to pin the schema to a specific version, you can configure the version in the URL.

```console
https://github.com/KeisukeYamashita/commitlint-rs/releases/download/v0.2.0/schema.json
```

In this case, the schema is pinned to version `0.2.0`.

:::

## Visual Studio Code

Configure your [Visual Studio Code](https://code.visualstudio.com/) to work with Commitlint.

### Edit in `settings.json`

Update your user `settings.json` or workspace settings (`/.vscode/settings.json`) to configure the schema.

#### JSON

```json
"json.schemas": [
  {
    "fileMatch": [
      ".commitlintrc",
      ".commitlintrc.json"
    ],
    "url": "https://github.com/KeisukeYamashita/commitlint-rs/releases/latest/download/schema.json"
  }
]
```

#### YAML

Associating schemas with YAMLs are supported by the [YAML language server](https://marketplace.visualstudio.com/items?itemName=redhat.vscode-yaml).

```json
"yaml.schemas": {
    "https://github.com/KeisukeYamashita/commitlint-rs/releases/latest/download/schema.json": [
        ".commitlintrc",
        ".commitlintrc.yaml",
        ".commitlint.yml"
    ]
}
```

### Specify schema in the configuration file

```json
{
  "$schema": "https://github.com/KeisukeYamashita/commitlint-rs/releases/download/v0.2.0/schema.json",
  "rules": {}
}
```

#### JSON

Add the following comment in your `.commitlintrc` or `.commitlintrc.json` file.

```json
{
  "$schema": "https://github.com/KeisukeYamashita/commitlint-rs/releases/latest/download/schema.json",
  "rules": {}
}
```

#### YAML

Associating schemas with YAMLs are supported by the [YAML language server](https://marketplace.visualstudio.com/items?itemName=redhat.vscode-yaml). Add the following comment in your `.commitlintrc`, `.commitlintrc.yaml` or `.commitlintrc.yml` file.

```yaml
# yaml-language-server: $schema=https://github.com/KeisukeYamashita/commitlint-rs/releases/latest/download/schema.json
rules:
```
