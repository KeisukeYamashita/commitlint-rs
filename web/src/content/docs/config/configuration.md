---
title: Configuration
description: Guide how to configure commitlint
---

Commitlint can be configured in many ways. You can use the default configuration or you can specify your own configuration file.

## Default configuration

If you don't specify any configuration file and you don't have any commitlint configuration files in your environment, the CLI will use the default configurations of each rules.

See the [default rules](/commitlint-rs/config/default) page for details.

## Location

Commitlint will look for configuration in the following places.

### Default

If no flag (see next section) is specified, the Commitlint will look for configuration in the following places **in the current working directory**:

* `.commitlintrc` (JSON or YAML file)
* `.commitlintrc.json` (JSON file)
* `.commitlintrc.yml` (YAML file)
* `.commitlintrc.yaml` (YAML file)

:::tip

Note that it is searched in the order written above and the first one found is loaded. Therefore, if you have `.commitlintrc` and `.commitlintrc.yml` in the same directory, the `.commitlintrc` will be loaded and the second one will be ignored.

:::

### Using the flag

Configuration file can be specified by using the `--config` flag or the short `-g` flag.

```console
# Using --config flag
$ commitlint --config path/to/.commitlintrc.yml

# Using -g flag
$ commitlint -g path/to/.commitlintrc.yml
```

If you specify a file and the file is not found, Commitlint will throw an error.

## Debug configuration

You can use the `--print-config` flag to print the configuration that will be used by Commitlint.

```console
$ commitlint --print-config
rules:
  description-empty: # Description must not be empty
    level: warning
  subject-empty: # Subject line must not be empty
    level: error
  type-empty: # Type must not be empty
    level: error
```
