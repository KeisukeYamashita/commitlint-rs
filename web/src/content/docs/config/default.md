---
title: Default Rules
description: List of the default rules
---

If you don't specify any configuration file and you don't have any commitlint configuration files in your environment, the CLI will use the default configurations of each rules describes in this page.

:::tip

You can also check the default values on the page of each rule.

:::

```yaml
rules:
    description-empty: # Description must not be empty
        level: warn
    scope-empty: # Scope must not be empty
        level: error
    subject-empty: # Subject line should exist
        level: error
    type-empty: # Type must not be empty
        level: error
```
