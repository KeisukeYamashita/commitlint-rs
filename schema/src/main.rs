use clap::Parser;
use cli::config::Config;
use std::fs;
/// CLI Arguments
#[derive(Parser, Debug)]
#[command(author, version, about, long_about = None)]
struct Args {
    /// Path to save the JSON schema
    #[arg(long, short)]
    path: String,
}

fn main() {
    let args = Args::parse();
    let config_schema = schemars::schema_for!(Config);
    let config_schema_json = serde_json::to_string_pretty(&config_schema).unwrap();
    fs::write(&args.path, config_schema_json).unwrap();
}
#[cfg(test)]
mod tests {
    #[test]
    fn snap_json_schema() {
        use cli::config::Config;
        use insta::assert_yaml_snapshot;
        let config_schema = schemars::schema_for!(Config);
        assert_yaml_snapshot!(config_schema);
    }
}
