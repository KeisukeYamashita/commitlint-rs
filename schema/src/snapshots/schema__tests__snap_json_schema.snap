---
source: schema/src/main.rs
expression: config_schema
snapshot_kind: text
---
$schema: "https://json-schema.org/draft/2020-12/schema"
title: Config
description: Config represents the configuration of commitlint.
type: object
properties:
  rules:
    description: Rules represents the rules of commitlint.
    $ref: "#/$defs/Rules"
required:
  - rules
$defs:
  BodyEmpty:
    description: "[BodyEmpty] represents the [`body-empty`](https://keisukeyamashita.github.io/commitlint-rs/rules/body-empty) rule."
    type: object
    properties:
      level:
        description: Level represents the level of the rule.
        anyOf:
          - $ref: "#/$defs/Level"
          - type: "null"
  BodyMaxLength:
    description: "[BodyMaxLength] represents the [`body-max-length`](https://keisukeyamashita.github.io/commitlint-rs/rules/body-max-length) rule."
    type: object
    properties:
      length:
        description: "Length represents the maximum length of the \"body\"."
        type: integer
        format: uint
        minimum: 0
      level:
        description: Level represents the level of the rule.
        anyOf:
          - $ref: "#/$defs/Level"
          - type: "null"
    required:
      - length
  DescriptionEmpty:
    description: "[DescriptionEmpty] represents the [`description-empty`](https://keisukeyamashita.github.io/commitlint-rs/rules/description-empty) rule."
    type: object
    properties:
      level:
        description: Level represents the level of the rule.
        anyOf:
          - $ref: "#/$defs/Level"
          - type: "null"
  DescriptionFormat:
    description: "[DescriptionFormat] represents the [`description-format`](https://keisukeyamashita.github.io/commitlint-rs/rules/description-format) rule."
    type: object
    properties:
      format:
        description: "Format represents the format of the \"description\"."
        type:
          - string
          - "null"
      level:
        description: Level represents the level of the rule.
        anyOf:
          - $ref: "#/$defs/Level"
          - type: "null"
  DescriptionMaxLength:
    description: "[DescriptionMaxLength] represents the [`description-max-length`](https://keisukeyamashita.github.io/commitlint-rs/rules/description-max-length) rule."
    type: object
    properties:
      length:
        description: "Length represents the maximum length of the \"description\"."
        type: integer
        format: uint
        minimum: 0
      level:
        description: Level represents the level of the rule.
        anyOf:
          - $ref: "#/$defs/Level"
          - type: "null"
    required:
      - length
  FootersEmpty:
    description: "[FootersEmpty] represents the [`footers-empty`](https://keisukeyamashita.github.io/commitlint-rs/rules/footers-empty) rule."
    type: object
    properties:
      level:
        description: Level represents the level of the rule.
        anyOf:
          - $ref: "#/$defs/Level"
          - type: "null"
  Level:
    description: Level represents the level of a rule.
    type: string
    enum:
      - error
      - ignore
      - warning
  Rules:
    description: "Rules represents the rules of commitlint.\n See: https://commitlint.js.org/reference/rules.html"
    type: object
    properties:
      type:
        anyOf:
          - $ref: "#/$defs/Type"
          - type: "null"
      body-empty:
        anyOf:
          - $ref: "#/$defs/BodyEmpty"
          - type: "null"
      body-max-length:
        anyOf:
          - $ref: "#/$defs/BodyMaxLength"
          - type: "null"
      description-empty:
        anyOf:
          - $ref: "#/$defs/DescriptionEmpty"
          - type: "null"
      description-format:
        anyOf:
          - $ref: "#/$defs/DescriptionFormat"
          - type: "null"
      description-max-length:
        anyOf:
          - $ref: "#/$defs/DescriptionMaxLength"
          - type: "null"
      footers-empty:
        anyOf:
          - $ref: "#/$defs/FootersEmpty"
          - type: "null"
      scope:
        anyOf:
          - $ref: "#/$defs/Scope"
          - type: "null"
      scope-empty:
        anyOf:
          - $ref: "#/$defs/ScopeEmpty"
          - type: "null"
      scope-format:
        anyOf:
          - $ref: "#/$defs/ScopeFormat"
          - type: "null"
      scope-max-length:
        anyOf:
          - $ref: "#/$defs/ScopeMaxLength"
          - type: "null"
      subject-empty:
        anyOf:
          - $ref: "#/$defs/SubjectEmpty"
          - type: "null"
      type-empty:
        anyOf:
          - $ref: "#/$defs/TypeEmpty"
          - type: "null"
      type-format:
        anyOf:
          - $ref: "#/$defs/TypeFormat"
          - type: "null"
      type-max-length:
        anyOf:
          - $ref: "#/$defs/TypeMaxLength"
          - type: "null"
  Scope:
    description: "[Scope] represents the [`scope`](https://keisukeyamashita.github.io/commitlint-rs/rules/scope) rule."
    type: object
    properties:
      level:
        description: Level represents the level of the rule.
        anyOf:
          - $ref: "#/$defs/Level"
          - type: "null"
      optional:
        description: "Optional scope. If true, even if the scope is not present, it is allowed."
        type: boolean
      options:
        description: "Options represents the options of the rule. If the option is empty, it means that no \"scope\" is allowed."
        type: array
        items:
          type: string
    required:
      - optional
      - options
  ScopeEmpty:
    description: "[ScopeEmpty] represents the [`scope-empty`](https://keisukeyamashita.github.io/commitlint-rs/rules/scope-empty) rule."
    type: object
    properties:
      level:
        description: Level represents the level of the rule.
        anyOf:
          - $ref: "#/$defs/Level"
          - type: "null"
  ScopeFormat:
    description: "[ScopeFormat] represents the [`scope-format`](https://keisukeyamashita.github.io/commitlint-rs/rules/scope-format) rule."
    type: object
    properties:
      format:
        description: "Format represents the format of the \"scope\"."
        type:
          - string
          - "null"
      level:
        description: Level represents the level of the rule.
        anyOf:
          - $ref: "#/$defs/Level"
          - type: "null"
  ScopeMaxLength:
    description: "[ScopeMaxLength] represents the [`scope-max-length`](https://keisukeyamashita.github.io/commitlint-rs/rules/scope-max-length) rule."
    type: object
    properties:
      length:
        description: "Length represents the maximum length of the \"scope\"."
        type: integer
        format: uint
        minimum: 0
      level:
        description: Level represents the level of the rule.
        anyOf:
          - $ref: "#/$defs/Level"
          - type: "null"
    required:
      - length
  SubjectEmpty:
    description: "[SubjectEmpty] represents the [`subject-empty`](https://keisukeyamashita.github.io/commitlint-rs/rules/subject-empty) rule."
    type: object
    properties:
      level:
        description: Level represents the level of the rule.
        anyOf:
          - $ref: "#/$defs/Level"
          - type: "null"
  Type:
    description: "[Type] represents the [`type`](https://keisukeyamashita.github.io/commitlint-rs/rules/type) rule."
    type: object
    properties:
      level:
        description: Level represents the level of the rule.
        anyOf:
          - $ref: "#/$defs/Level"
          - type: "null"
      options:
        description: "Options represents the options of the rule. If the option is empty, it means that no \"type\" is allowed."
        type: array
        items:
          type: string
    required:
      - options
  TypeEmpty:
    description: "[TypeEmpty] represents the [`type-empty`](https://keisukeyamashita.github.io/commitlint-rs/rules/type-empty) rule."
    type: object
    properties:
      level:
        description: Level represents the level of the rule.
        anyOf:
          - $ref: "#/$defs/Level"
          - type: "null"
  TypeFormat:
    description: "[TypeFormat] represents the [`type-format`](https://keisukeyamashita.github.io/commitlint-rs/rules/type-format) rule."
    type: object
    properties:
      format:
        description: "Format represents the format of the \"type\"."
        type:
          - string
          - "null"
      level:
        description: Level represents the level of the rule.
        anyOf:
          - $ref: "#/$defs/Level"
          - type: "null"
  TypeMaxLength:
    description: "[TypeMaxLength] represents the [`type-max-length`](https://keisukeyamashita.github.io/commitlint-rs/rules/type-max-length) rule."
    type: object
    properties:
      length:
        description: "Length represents the maximum length of the \"type\"."
        type: integer
        format: uint
        minimum: 0
      level:
        description: Level represents the level of the rule.
        anyOf:
          - $ref: "#/$defs/Level"
          - type: "null"
    required:
      - length
